design/ceph-deprecate-old-osd-types.md

```diff
@@ -0,0 +1,433 @@
+# Ceph OSD redesign for Ceph Octopus (May 2019)
+**Targeted for v1.1-v1.2**
+
+## Background
+Rook's current implementation for preparing and starting OSDs includes several different patterns
+that have evolved over Rook's lifetime. These patterns include different code paths for supporting
+directories and devices as storage targets, and further separated into support for both Filestore
+and Bluestore as backing stores. There is further complication added by support for `ceph-volume` as
+a tool for storage provisioning versus continued support for legacy installations of Rook which used
+manual partitioning.
+
+There has been talk for quite some time about removing support for directories as storage targets,
+as this is a use case in Ceph that is not supported for production. Additionally, Filestore is
+effectively in sustaining mode, and Bluestore is the defacto replacement for it.
+
+All of the various OSD support patterns/code-paths need changes to allow the [planned updates to
+Rook's configuration strategy](/design/ceph-config-updates.md) to proceed. The various code paths
+are intertangled and difficult to separate, and this presents a huge hurdle for the config changes.
+In general, the intertangled OSD code poses a risk for long-term maintenance, as it may be more
+prone to bugs being introduced, and it has and will continue to slow development.
+
+Atop all of this, there have been discussions ongoing about supporting Ceph's Rook
+orchestrator as the intended way of managing Ceph deployments on Kubernetes. Orchestration currently
+modifies the Ceph cluster resource, but from the standpoint of a Ceph storage administrator, the CRD
+(and Rook) lacks the flexibility needed for non-trivial Ceph deployments. While flexibility could be
+added, it will be better to remove the need for such complexity in the cluster CRD and in Rook. The
+details of this part of the design will be expounded later.
+
+In light of this background information, this design proposes three high-level goals:
+1. Identify OSD features/code-paths that should be deprecated due to having become outmoded in some
+   form.
+2. Propose a new architecture for Rook's OSD code that is compatible with the planned config updates
+   and that is extensible for the future.
+3. Provide a timeline for deprecating outmoded, and provide users tools for transitioning their
+   deployments which use old methods to begin using the new methods.
+
+
+## Deprecated features and code paths
+This design proposes to deprecate ...
+- the use of directories as storage targets
```

Comment:
Hm, that's not very good for small setups, where you don't have separated drives to format. Use of directories as storage targets was main argument for me for choosing Rook over OpenEBS. Using directories also helps utilizing OS partition storage, which with some setups might be significant (think bare metal server with 2TB in RAID 1).

Reply:
This proposal introduces a replacement for this idea of installing bluestore devices in files. Take a look and let us know if you see that as being feasible.

The reasoning behind this is that supporting directories as storage targets using Ceph is a kind of degraded use-case that to my knowledge isn't supported by Ceph. My understanding is that the feature exists in Ceph primarily for test cases. Even installing bluestore in a file isn't really "supported" but exists for testing scenarios. My concern is that Rook could be relying on unsupported behavior in Ceph for a supported (or seemingly supported) use-case, which means we may be putting user data at unnecessary risk without having users able to make informed decisions for their data. I also believe that using directories as OSDs risks easily running out of space on the OS disk. Relatedly, Ceph's usage reporting probably isn't accurate in this case since Ceph isn't designed for this use-case.

Something we could also discuss here is talking about support for using partitions as osds, which would be a safer solution for user data and OS data. I'm not sure if there are Ceph concerns related to partition use.

Reply:
Yeah, `directories:` are still needed for the future at least for testing purposes only as a good amount of people are using it.

Reply:
Again, this proposal does have the test use case in mind. For testing purposes, we are proposing using a preallocated file as the bluestore target.

Reply:
Thank you for your explanation @BlaineEXE.

While preallocated file option is not ideal, as you lose the flexibility of storage (you need to decide how much space you dedicate for ceph upfront), it's better than requiring store device (maybe loop can be used to skip preallocated file?)

I understand that this decision is driven by upstream Ceph and the proposal absolutely make sense in this context.

Reply:
Could a production-supportable solution be to allow partitions as storage targets? Can `ceph-volume` use partitions? @leseb @jan--f

Reply:
@BlaineEXE `ceph-volume` can essentially create OSDs on any LV. If its acceptable for smaller deployments or testing to create the wanted LVs manually, the underlying storage is up to the user. This might however make the specification of which storage to use more complex.